src/java/genomicsdb/VCF2TileDB.java

```diff
@@ -75,8 +368,316 @@ public VCF2TileDB(String loaderJSONFile)
+    {
+        if(mIsLoaderSetupDone)
+            return;
+        //Callset mapping JSON - convert to string
+        JSONObject topCallsetJSON = new JSONObject();
+        topCallsetJSON.put("callsets", mCallsetMappingJSON);
+        StringWriter stringWriter = new StringWriter();
+        topCallsetJSON.writeJSONString(stringWriter);
+        //Call native setupGenomicsDBImporter()
+        mMaxBufferStreamIdentifiers = jniSetupGenomicsDBLoader(mGenomicsDBImporterObjectHandle, stringWriter.toString());
+        //Why 2* - each identifier is a pair<buffer_stream_idx, partition_idx>
+        //Why +1 - the last element will contain the number of exhausted stream identifiers when importBatch() is called
+        mExhaustedBufferStreamIdentifiers = new long[2*((int)mMaxBufferStreamIdentifiers)+1];
+        //Set all streams to empty
+        //Add all streams to mExhaustedBufferStreamIdentifiers - this way when importBatch is called the first
+        //time all streams' data are written
+        for(int i=0,idx=0;i<mBufferStreamWrapperVector.size();++i,idx+=2)
+        {
+            SilentByteBufferStream currStream = mBufferStreamWrapperVector.get(i).mStream;
+            currStream.setNumValidBytes(0);
+            mExhaustedBufferStreamIdentifiers[idx] = i;
+            mExhaustedBufferStreamIdentifiers[idx+1] = 0;
+        }
+        //Set number of exhausted buffer streams - all streams are exhausted the first time
+        mNumExhaustedBufferStreams = mBufferStreamWrapperVector.size();
+        mExhaustedBufferStreamIdentifiers[(int)(2*mMaxBufferStreamIdentifiers)] = mNumExhaustedBufferStreams;
+        mIsLoaderSetupDone = true;
+    }
+
+    /**
+     * Write VariantContext object to stream - may fail if the buffer is full
+     * It's the caller's responsibility keep track of the VC object that's not written
+     * @param vc VariantContext object
+     * @param streamIdx index of the stream returned by the addBufferStream() call
+     * @return true if the vc object was written successfully, false otherwise
+     */
+    public boolean add(VariantContext vc, final int streamIdx) throws GenomicsDBException, RuntimeIOException
+    {
+        if(streamIdx < 0 || streamIdx >= mBufferStreamWrapperVector.size())
+            throw new GenomicsDBException("Invalid stream idx "+Integer.toString(streamIdx)+" must be between [0-"
+                    +Long.toString(mBufferStreamWrapperVector.size()-1)+"]");
+        if(!mIsLoaderSetupDone)
+            throw new GenomicsDBException("Cannot add VariantContext objects to streams before calling setupGenomicsDBImporter()");
+        GenomicsDBImporterStreamWrapper currWrapper = mBufferStreamWrapperVector.get(streamIdx);
+        currWrapper.mVCWriter.add(vc);
+        SilentByteBufferStream currStream = currWrapper.mStream;
+        if(currStream.overflow() && currStream.getMarker() > 0) //at least one record already existed in the buffer
```

Comment:
Will it simplify code if we handle the overflow condition inside SilentByteBuffeStream write() method?